onnxmltools/utils/metadata_props.py

```diff
@@ -0,0 +1,82 @@
+    'Image.ColorSpaceGamma': ['linear', 'srgb'],
+    'Image.NominalPixelRange': ['nominalrange_0_255', 'normalized_0_1', 'normalized_1_1', 'nominalrange_16_235'],
+})
+
+
+def _validate_metadata(metadata_props):
+    '''
+    Validate metadata properties and possibly show warnings or throw exceptions.
+
+    :param metadata_props: A dictionary of metadata properties, with property names and values (see :func:`~onnxmltools.utils.metadata_props.add_metadata_props` for examples)
+    '''
+    if len(CaseInsensitiveDict(metadata_props)) != len(metadata_props):
+        raise RuntimeError('Duplicate metadata props found')
+
+    for key, value in metadata_props.items():
+        valid_values = KNOWN_METADATA_PROPS.get(key)
+        if valid_values and value.lower() not in valid_values:
+            warnings.warn('Key {} has invalid value {}. Valid values are {}'.format(key, value, valid_values))
+
+
+def add_metadata_props(onnx_model, metadata_props, targeted_onnx=onnx.__version__):
+    '''
+    Add metadata properties to the model. See recommended key names at:
+    `Extensibility - Metadata <https://github.com/onnx/onnx/blob/296953db87b79c0137c5d9c1a8f26dfaa2495afc/docs/IR.md#metadata>`_ and
+    `Optional Metadata <https://github.com/onnx/onnx/blob/master/docs/IR.md#optional-metadata>`_
+
+
+    :param onnx_model: ONNX model object
+    :param metadata_props: A dictionary of metadata properties, with property names and values, such as { 'model_author': 'Alice', 'model_license': 'MIT' }
+    :param targeted_onnx: Target ONNX version
+    '''
+    if StrictVersion(targeted_onnx) < StrictVersion('1.2.1'):
+        warnings.warn('Metadata properties are not supported in targeted ONNX-%s' % targeted_onnx)
+        return
+    _validate_metadata(metadata_props)
+    new_metadata = CaseInsensitiveDict({x.key: x.value for x in onnx_model.metadata_props})
+    new_metadata.update(metadata_props)
+    del onnx_model.metadata_props[:]
+    onnx_model.metadata_props.extend(
+        onnx_proto.StringStringEntryProto(key=key, value=value)
+        for key, value in metadata_props.items()
+    )
+
+
+def set_denotation(onnx_model, input_name, denotation, dimension_denotation=None, targeted_onnx=onnx.__version__):
+    '''
+    Set input type denotation and dimension denotation.
+
+    Type denotation is a feature in ONNX 1.2.1 that let's the model specify the content of a tensor (e.g. IMAGE or AUDIO).
+    This information can be used by the backend. One example where it is useful is in images: Whenever data is bound to
+    a tensor with type denotation IMAGE, the backend can process the data (such as transforming the color space and
+    pixel format) based on model metadata properties.
+
+    :param onnx_model: ONNX model object
+    :param input_name: Name of input tensor to edit
+    :param denotation: Input type denotation (`documentation <https://github.com/onnx/onnx/blob/master/docs/TypeDenotation.md#type-denotation-definition>`_)
```

Comment:
What's its type? To know that it's a string, one needs to read [this file](https://github.com/onnx/onnx/blob/7848f1e0414ba3b2e263609d93d46fd60790b2e9/onnx/onnx.proto#L402) and trace the code below. However, for a new user, it could be very painful to figure out what this attribute should look like. Could you please specify all input types and add an example per attribute for all user-facing functions? I know you have a test but users are unlikely going to read it. Many thanks.

Reply:
It it better now?

Reply:
Yes, so only one last comment left; please copy the metadata from duplicate to original (as described in my general comment in my last review).